### PR TITLE
fmt: fix topic normalisation

### PR DESCRIPTION
**Pull request by robphoenix**
See https://github.com/exercism/meta/issues/83

### CODE REVIEW

File: cmd/fmt.go [hunk 1]
```diff
@@ -145,7 +146,19 @@ func formatTopics(m map[string]interface{}) {
...
 func normaliseTopic(t string) string {
 	s := strings.ToLower(t)
-	s = strings.Replace(s, " ", "-", -1)
+
+	// we only want letters, hyphens and
+	// whitespace (to be replaced) in topics
+	reg := regexp.MustCompile("[^a-zA-Z\\s-]+")
+	s = reg.ReplaceAllString(s, "")
+
+	// corner case where prefix should be removed
```
Comment (robphoenix) [inline cmd/fmt.go]: This hasn't been confirmed yet, but I thought I'd put it in here for now, just wanted to consider how it would be done.

Comment (kytrinyx) [reply cmd/fmt.go]: Sounds good. I'll try to push all of this to be confirmed today.

File: cmd/fmt.go [hunk 2]
```diff
@@ -145,7 +146,17 @@ func formatTopics(m map[string]interface{}) {
 
 func normaliseTopic(t string) string {
 	s := strings.ToLower(t)
-	s = strings.Replace(s, " ", "-", -1)
+
+	// we only want to let through letters and underscores
+	// for the final output.
+	// hyphens and whitespace are allowed through for now
+	// as word delimiters to be replaced with underscores.
+	reg := regexp.MustCompile("[^a-zA-Z\\s-_]+")
```
Comment (nywilken) [inline cmd/fmt.go]: There's a call to ToLower above so the regex could be simplified to `[^a-z\\s-_]+`. Or even `(?i)[^a-z\\s-_]+` if you prefer the regex to be case insensitive.

File: cmd/fmt.go [hunk 3]
```diff
@@ -145,7 +146,17 @@ func formatTopics(m map[string]interface{}) {
...
+	s = reg.ReplaceAllString(s, "")
+
+	// output to be snake_case
+	s = strings.Replace(s, " ", "_", -1)
```
Comment (nywilken) [inline cmd/fmt.go]: If there is a topic that has a double space or double dash (user typo) the resulting string will have a double underscore. What do you think about using another regex replace here with the pattern `[\\s-]+`?

**Review by nywilken — COMMENTED** [review]
@robphoenix this looks good - just a few suggestions  on the string replacements.

File: cmd/fmt.go (hunk 2 above)
Comment (robphoenix) [reply cmd/fmt.go]: ah yes, of course, thanks!

File: cmd/fmt.go (hunk 3 above)
Comment (robphoenix) [reply cmd/fmt.go]: This is a much nicer way of doing it, avoiding multiple calls and covering more cases. :pray:

File: fixtures/format/malformed/config.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
       "topics": [
 "equality",
              "mathematics",
-                    "text-formatting",
+                    "text--formatting",
```
Comment (robphoenix) [inline fixtures/format/malformed/config.json:18]: I updated the malformed test case to reflect the improvements suggested by @nywilken

**Review by nywilken — APPROVED** [review]

**Comment by nywilken** [comment]
@kytrinyx @robphoenix this looks good and I would like to merge it if there are no any pending items being discussed?

**Comment by robphoenix** [comment]
It's good to go from my point of view.

**Review by kytrinyx — APPROVED** [review]
Let's ship this!

**Comment by nywilken** [comment]
@robphoenix I missed your response. Thanks for merging.

**Pull request merged**